Clean up headers

veblush opened the pull request:
Clean up public header files not to include `<mutex>` and `<condition_variable>` because i) it's not used and ii) discouraged to use in gRPC libraries.

vjpai commented on include/grpcpp/impl/codegen/sync.h:
```diff
@@ -56,13 +54,7 @@ class Mutex {
-#endif
-  };
+  gpr_mu mu_;
 };
```
This one is a special case. We have to keep this one. We need a union that is as big as the bigger of gpr_mu, std::mutex, and pthread_mutex_t in order to be able to run multi-platform. This has caused problems for us in the past.

veblush replied:
Can you elaborate this a bit more? It isn't clear to me why it's required. Does gRPC support link-time multi-platform such as building codegen with pthread and linking it with something with pthread-alternative (like absl::mutex)?

vjpai replied:
Chat with @soheilhy about this topic.

veblush replied:
I did actually and it seems that there was ODR violation and it required this hack to work it around. But we're not sure it's still required. I'll try to remove this later.

soheilhy replied:
Thank you. It'd be great if we can make sure this never happens and then get rid of this hack!

veblush commented:
Thanks for the review. I added additional check for #include to `cpp_banned_constructs.sh` so that disallowed #include cannot get in. To make this PR get in quickly, I removed the change of sync.h.

vjpai commented on tools/run_tests/sanity/cpp_banned_constructs.sh:
```diff
@@ -29,3 +29,12 @@ egrep -Irn \
+
+egrep -Irn \
+    '^#include (<mutex>|<condition_variable>|<thread>|<ratio>|<filesystem>|<future>|<system_error>)' \
+    include/grpc include/grpcpp src/core | \
```
Why not src/cpp also?

veblush replied:
Oops my bad. Fixed it.

vjpai commented:
Thanks for doing this and finding this!

jtattermusch commented:
@veblush @rmstar  I think this broke the `flaky_network_test` on master.

https://source.cloud.google.com/results/invocations/4ebd7fa4-4d05-430f-a477-198abe91d2ce/log
```
ERROR: /var/local/git/grpc/test/cpp/end2end/BUILD:649:1: C++ compilation of rule '//test/cpp/end2end:flaky_network_test' failed (Exit 1)
test/cpp/end2end/flaky_network_test.cc:264:21: error: no type named 'condition_variable' in namespace 'std'
               std::condition_variable* cond) {
               ~~~~~^
test/cpp/end2end/flaky_network_test.cc:255:12: error: no type named 'condition_variable' in namespace 'std'
      std::condition_variable cond;
      ~~~~~^
2 errors generated.
```
